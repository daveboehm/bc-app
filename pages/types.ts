// --------------
// Private types
// --------------

type ImageType = {
    height: number;
    img?: HTMLElement;
    title: string;
    url: string;
    width: number;
}

type UrlType = {
    value: string;
    // @TODO: confirm available options for link target attr
    target: "_blank" | "parent" | "self";
}

// --------------
// External types
// --------------

export type EntryType = {
    description?: string;
    id: number;
    image: ImageType;
    title: string;
    url: UrlType;
}

export type EntriesPropsType = {
    entryIds: (string | undefined)[];
    hydratedEntries: (EntryType | undefined)[];
}

export type PaginationProps = {
    currentPage: number;
    itemsPerPage?: number;
    handlePageChange: (_pageNumber: number) => void;
    totalItems: number;
}