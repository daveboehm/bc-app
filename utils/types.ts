// --------------
// Private types
// --------------

type ImageType = {
    height: number;
    img?: HTMLElement;
    title: string;
    url: string;
    width: number;
}

type UrlType = {
    value: string;
    target: "_blank" | "_parent" | "_self" | "_top";
}

// --------------
// External types
// --------------

export type EntryType = {
    description?: string;
    id: number;
    image: ImageType;
    title: string;
    url: UrlType;
}

export type EntriesPropsType = {
    entryIds: (string | undefined)[];
    hydratedEntries: (EntryType | undefined)[];
}

export type PaginationProps = {
    currentPage: number;
    itemsPerPage?: number;
    handlePageChange: (_pageNumber: number) => void;
    totalItems: number;
}