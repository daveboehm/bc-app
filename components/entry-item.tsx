import React from "react"
import type { EntryType } from "../utils/types"
import styles from '../styles/Entry.module.css'
import Image from "next/image";

export const EntryItem = (props: EntryType) => {
    const { id, image, title, url } = props;

    if (!url || !id || !title) return null;

    return (
        <li className={styles.entryItem} data-testid="entry-item">
            <a className={styles.entryItemLink} href={url.value} target={url.target} title={image.url}>
                <div className={styles.entryImage}>
                    <Image src={image.url} alt={image.title} width={image.width} height={image.height} />
                </div>
                <h2 className={styles.entryTitle}>{title}</h2>
            </a>
        </li>
    )
}

export default EntryItem