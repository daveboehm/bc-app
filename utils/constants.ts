export const ENTRIES_URL = 'https://www.bigcommerce.com/actions/bcCore/interview/getShowcaseEntryIds'

export const ENTRY_URL = 'https://www.bigcommerce.com/actions/bcCore/interview/getShowcaseEntryById'

export const ITEMS_PER_PAGE = 10