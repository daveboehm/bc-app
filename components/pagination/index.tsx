import React, { useState, useEffect } from "react"
import type { PaginationProps } from "../../utils/types"
import { ITEMS_PER_PAGE } from "../../utils/constants"
import styles from '../styles/Pagination.module.css'

export const Pagination = (props: PaginationProps) => {
    const {
        currentPage,
        handlePageChange,
        itemsPerPage = ITEMS_PER_PAGE,
        totalItems
    } = props

    const [pageButtons, setPageButtons] = useState(1)

    useEffect(() => {
        const calcPageButtons = Math.ceil(totalItems / itemsPerPage)
        setPageButtons(calcPageButtons)
    }, [totalItems, itemsPerPage])

    return (
        <div className={styles.paginationContainer}>
            <div className={styles.buttonsContainer}>

                <button
                    data-testid="pagination-prev-button"
                    disabled={currentPage <= 0}
                    type="button"
                    className={styles.pageButtonInactive}
                    onClick={() =>handlePageChange(currentPage - 1)}>Prev</button>
                
                {pageButtons > 1 && [...Array(pageButtons)].map((_undef, pageIndex) => (
                    <button
                        data-testid="pagination-button"
                        type="button"
                        key={pageIndex}
                        className={pageIndex === currentPage ? styles.pageButtonActive : styles.pageButtonInactive}
                        onClick={() => handlePageChange(pageIndex)}>
                        {pageIndex + 1}
                    </button>
                    ))}

                <button
                    data-testid="pagination-next-button"
                    disabled={currentPage >= pageButtons-1}
                    type="button"
                    className={styles.pageButtonInactive}
                    onClick={() => handlePageChange(currentPage + 1)}>Next</button>
                
            </div>
        </div>
    )
}

export default Pagination