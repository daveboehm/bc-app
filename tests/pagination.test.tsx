import React from 'react'
import { cleanup, render, screen, RenderResult } from "@testing-library/react";
import { PaginationProps } from '../utils/types';
import { MOCK_PAGINATION_PROPS } from './__fixtures__/mock-data';
import { Pagination } from '../components/pagination';

describe("<Pagination />", () => {

    describe("happy path", () => {
        let props: PaginationProps
        let view: RenderResult

        beforeEach(() => {
            props = MOCK_PAGINATION_PROPS
            view = render(<Pagination {...props} />)
        })

        afterEach(cleanup)

        it("should render and match snapshot", () => {
            expect(view.container).toBeVisible()
            expect(view.container).toMatchSnapshot()
        })

        it("should display 3 pages possible", () => {
            const pageButtons = screen.getAllByTestId(/pagination-button/)
            expect(pageButtons).toHaveLength(3) // @Note: Mock data has 25 total items, showing 10 per page
        })
    })
    
})