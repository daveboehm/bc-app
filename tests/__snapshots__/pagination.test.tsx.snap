// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> happy path should render and match snapshot 1`] = `
<div>
  <div
    class="paginationContainer"
  >
    <div
      class="buttonsContainer"
    >
      <button
        class="pageButtonInactive"
        data-testid="pagination-prev-button"
        type="button"
      >
        Prev
      </button>
      <button
        class="pageButtonInactive"
        data-testid="pagination-button"
        type="button"
      >
        1
      </button>
      <button
        class="pageButtonInactive"
        data-testid="pagination-button"
        type="button"
      >
        2
      </button>
      <button
        class="pageButtonActive"
        data-testid="pagination-button"
        type="button"
      >
        3
      </button>
      <button
        class="pageButtonInactive"
        data-testid="pagination-next-button"
        disabled=""
        type="button"
      >
        Next
      </button>
    </div>
  </div>
</div>
`;
